server/graphql/resolvers.ts

```diff
@@ -35,9 +35,11 @@ const resolvers = {
     requestTypes: (parent, __, { dataSources }): Array<RequestTypeInterface> => parent.requestTypes.map(id => dataSources.requestTypes.getById(Types.ObjectId(id)))
   },
   RequestType: {
-    requestGroup: (parent, __, { dataSources }): RequestGroupInterface => dataSources.requestGroups.getById(Types.ObjectId(parent.requestGroup.toString())),
-    numOpen: (parent, __, { dataSources }): Number => parent.requests.open.length,
-    requests: (parent, __, { dataSources }): Array<RequestInterface> => parent.requests.map(id => dataSources.requests.getById(Types.ObjectId(id)))
+    numOpen: (parent, __, { dataSources }): Number => parent.requests.map(id => dataSources.requests.getById(id)).reduce((total, request) => (request.deleted === false && request.fulfilled === false) ? total + 1 : total, 0),
+    openRequests: (parent, __, { dataSources }): Array<RequestInterface> => parent.requests.map(id => dataSources.requests.getById(id)).filter(request => request.fulfilled === false && request.deleted === false),
```

Comment:
Since `numOpen` is basically filtering open requests, the logic is (kinda) duplicated with `openRequests`

I recommend extracting out the logic for filtering requests. Feel free to make a new file for the functions
```
const resolvers = {
 ...
  RequestType: {
    numOpen: (parent, __, { dataSources }): filterOpenRequests(getRequestsById(parent.requests)).length,
    openRequests: (parent, __, { dataSources }): Array<RequestInterface> => filterOpenRequests(getRequestsById(parent.requests)),
    fulfilledRequests: (parent, __, { dataSources }): Array<RequestInterface> => filterFulfilledRequests(getRequestsById(parent.requests)),
    deletedRequests: (parent, __, { dataSources }): Array<RequestInterface> => filterDeletedRequests(getRequestsById(parent.requests)),
  }
  ...
}

const filterOpenRequests = (requests: Array<RequestInterface>){
  return requests.filter(request => request.fulfilled === false && request.deleted === false)
}

const filterFulfilledRequests = (requests: Array<RequestInterface>){
  return requests.filter(request => request.fulfilled === true && request.deleted === false)
}

const filterDeletedRequests = (requests: Array<RequestInterface>){
  return requests.filter(request => request.deleted === true)
}

const getRequestsById = (requestIds, dataSources) => {
  return requestIds.map(id => dataSources.requests.getById(id))
}
```

Going one step further, we could add a `getByIds` function to `CachedMongooseDataSource`.

But honestly you can choose to not do this or  make it a separate PR
```
getByIds(ids: Array<Types.ObjectId>): Array<DocumentType> {
    return ids.map((id) => this.getById(id))
  }
```

Reply:
^ I think we are seeing more and more uses of getting an array of IDs with numOpen, etc. It may simplify some resolvers. It may be worthwhile going forward with this

Reply:
I agree. It's a pain to manage the 3 arrays when you update stuff. I like this approach a lot more.